[AArch64] Optimize extending loads of small vectors

guy-david opened the pull request:
Reduces the total amount of loads and the amount of moves between SIMD registers and general-purpose registers.

davemgreen reviewed (commented):
4 x i8 is custom lowered. Could we do the same thing here, or does that not work as well?

guy-david committed:
[AArch64] Optimize extending loads of small vectors

Reduces the total amount of loads and the amount of moves between SIMD
registers and general-purpose registers.

davemgreen commented:
Sorry for the delay - I was away for a while. Do you mean you tried it and it didn't work very well? I can imagine that it might be better to keep the extend separate from the load.

The v4i8 lowering was added in https://reviews.llvm.org/D105110. I don't know if there is a strong reason to pick lowering vs a combine that happens earlier, but we should presumably have one code path that works for all of the types. Is it possible to make them work the same way, either via a combine or all via lowering?

davemgreen commented:
> This patch removes the custom lowering for v4i8

This current patch? It seems to just add a combine for `MemVT == MVT::v2i8 || MemVT == MVT::v2i16` loads? I might be missing something, let me know.

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -7188,12 +7230,89 @@ SDValue AArch64TargetLowering::LowerStore128(SDValue Op,
+      DAG.getNode(ISD::SCALAR_TO_VECTOR, DL, ScalarToVecTy, ScalarLoad);
+  MVT BitcastTy =
+      MVT::getVectorVT(MVT::getIntegerVT(SrcEltBits), 128 / SrcEltBits);
+  SDValue Bitcast = DAG.getNode(ISD::BITCAST, DL, BitcastTy, ScalarToVec);
```
Does it need to be careful about BE? (Is it OK because the load and the bitcast cancel out?)

davemgreen reviewed (commented):
Thanks - this looks decent.

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -1427,12 +1427,27 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
+    setLoadExtAction(ISD::ZEXTLOAD, MVT::v2i64, MVT::v2i16, Custom);
+    setLoadExtAction(ISD::EXTLOAD, MVT::v4i32, MVT::v4i16, Custom);
+    setLoadExtAction(ISD::SEXTLOAD, MVT::v4i32, MVT::v4i16, Custom);
+    setLoadExtAction(ISD::ZEXTLOAD, MVT::v4i32, MVT::v4i16, Custom);
```
A v4i16 vector load is simple and legal, and the extend is a legal extend so I wouldn't expect it to go via custom lowering of an extending-load. Does it help to remove v4i16 from isEligibleForSmallVectorLoadOpt?

guy-david committed:
Addressed comments

davemgreen reviewed (approved):
Thanks. LGTM
